src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -45,71 +114,131 @@ ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer,
+        };
 
-        if (opcode == spv::OpEntryPoint)
+        if (entryPoint.shaderStage == hlsl::ESS_UNKNOWN)
         {
-            const auto curExecutionModel = static_cast<spv::ExecutionModel>(spirv[curOffset + 1]);
-
-            // TODO: check whether this reinterpret_cast is UB
-            const auto curEntryPointName = std::string_view(reinterpret_cast<const char*>(spirv + curOffset + 3)); // entryPoint name is the third parameter
+            logger.log("Found entry point with unsupported execution model in SPIR-V", system::ILogger::ELL_ERROR);
+            return Result{
+                .spirv = nullptr,
+                .isSuccess = false
+            };
+        }
 
-            auto findEntryPointIt = std::find(entryPoints.begin(), entryPoints.end(), 
-              EntryPoint{
-                .name = curEntryPointName,
-                .shaderStage = getHlslShaderStage(curExecutionModel),
-              });
-            if (findEntryPointIt != entryPoints.end())
-            {
-                foundEntryPoints[std::distance(findEntryPointIt, entryPoints.begin())] = true;
-            } else
+        auto findEntryPointIt = unfoundEntryPoints.find(entryPoint);
+        if (findEntryPointIt != unfoundEntryPoints.end())
+        {
+            unfoundEntryPoints.erase(findEntryPointIt);
+        } else
+        {
+            if (needDebloat == false)
             {
-                if (needDebloat == false)
-                {
-                    minimizedSpirv.reserve(spirvDwordCount);
-                    minimizedSpirv.insert(minimizedSpirv.end(), spirv, spirv + curOffset);
-                    needDebloat = true;
-                }
-                continue;
+                minimizedSpirv.reserve(spirvDwordCount);
+                minimizedSpirv.insert(minimizedSpirv.end(), spirv, spirv + curOffset);
+                needDebloat = true;
             }
+            removedEntryPointIds.insert(curEntryPointId);
+            continue;
         }
-        
         if (!needDebloat) continue;
         minimizedSpirv.insert(minimizedSpirv.end(), spirv + curOffset, spirv + offset);
     }
 
-    const auto isAllEntryPointsFound = std::all_of(foundEntryPoints.begin(), foundEntryPoints.end(), std::identity{});
+    const auto wereAllEntryPointsFound = unfoundEntryPoints.empty();
+    if (!wereAllEntryPointsFound)
+    {
+        logger.log("Some entry point that is requested to be retained is not found in SPIR-V", system::ILogger::ELL_ERROR);
+        return {
+            .spirv = nullptr,
+            .isSuccess = false,
+        };
+    }
 
-    if (!isAllEntryPointsFound || !needDebloat) {
-        return Result {
-            .isAllEntryPointsFound = isAllEntryPointsFound,
+    if (!needDebloat)
+    {
+        return {
             .spirv = nullptr,
+            .isSuccess = true,
         };
     }
 
+    // handle execution model removal
+    while (offset < spirvDwordCount)
```

Comment:
quote https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#LogicalLayout in the comment so not resetting `offset` is not taken to be a bug